### ISSUE TITLE
[json] use the schemastore catalog

### ISSUE THREAD

**Issue by aeschli** [Microsoft/vscode#26289]
The JSON server should use http://schemastore.org/api/json/catalog.json for schema associations

**Comment by DoCode** [comment]
@aeschli, what is the status here?
How can we use a schema catalog with VS Code?

**Comment by aeschli** [comment]
Schema associations can be set in the user settings and and by extensions. But we're not defaulting to the http://schemastore.org/api/json/catalog.json settings yet. No progress yet.

**Comment by DoCode** [comment]
@aeschli, thanks.  
But my question was, how a schema catalog(catalog.json) can be used, like in VS.  
Currently, I see the only option is, that the user must set a schema explicit per file.  Schemastore.org has an elegant solution here.

**Comment by felixfbecker** [comment]
After having such a great experience with most JSON files, and using this feature with the YAML language server, I was just wondering why a certain schema I added to SchemaStore wouldn't be used automatically without `$schema`. This would be really great to have

**Comment by jabby** [comment]
Hello @aeschli 
As you can see in this pull request https://github.com/eclipse/wildwebdeveloper/pull/130, the feature will be really interesting for multiple editor. 

I would like to contribute a patch based on the work already done on yaml-language-server project here https://github.com/redhat-developer/yaml-language-server/blob/master/src/server.ts#L269-L302

First, are you ok with my proposal. If so how can I test my work. I don't see any information on how launch the server and connect my vscode on it.

**Comment by jabby** [comment]
In fact, I should read the main README of this repository. I will come back soon.

**Comment by aeschli** [comment]
Yes, see https://github.com/Microsoft/vscode/blob/master/extensions/json-language-features/CONTRIBUTING.md

Yes, help is welcome, it would be great if you can give it a try.
What's important is that schema associations that are made in the user settings or by extensions will win against the associations from the catalog.

I'm worried that there will be wrong associations to certain files with common file names. That will lead to misleading code completions, hovers, and even validation errors. We need an easy way for users to disable automatic associations on such files.
I'm not really sure how we can do this. Any ideas?

**Comment by jabby** [comment]
I give a try this week. I hope I will find time during lunch time. 

I agree with you about the user settings. There are more important than automatic settings from the catalog. However I don't understand the extensions part of your message. Which extensions are you refering to?

For wrong association, I think we can create a settings like in the yaml-language-server. 

```json
"json": {
            // other settings like format and schemas associations
            "schemaStore": {
                "enable": true 
            }
        }
```

If the schemastore is disabled, we don't use the catalog to create associations between files and schemas. This mecanism can also be extended by another key in `schemastore` setting. I think a property like `disableAssociations` as an array of fileMatch is a quite good idea. What do you think about this?

**Comment by bd82** [comment]
Should installed Extensions exposing: [contributes.jsonValidation](https://code.visualstudio.com/api/references/contribution-points#contributes.jsonValidation) have precedence on schemas coming from the SchemaStore?

**Comment by aeschli** [comment]
Yes, schemas from the schema store should only be used if there's no association (as a fallback)

**Comment by csvn** [comment]
So many times schemas seem to "just work"™, that it's not really obvious at times why it does not. It's hard to realize that they are often provided by extensions. I just got hit by this issue when I was confused by `renovate.json` not loading the schema correctly.

Is there any work done here currently? #69132 seems to have stalled 😥

**Comment by conkman** [comment]
Is this still a issue?  We are about to publish a schema to Schema Store and VSCode is the first IDE we want to beta test with.

**Comment by baywet** [comment]
Hey everyone,
While I'm not on the VSCode team, and have no control over when the pull request will be merged, I put together #327628 to implement support for schema store.
Between that, the recently added support for JSON schema 2019-09 and 2020-12, and another bugfix I've worked on, the end to end story of "I open a JSON document, and it magically find its corresponding schema to give me structural validation" is working well locally. 

Now we just need to get it merged/released :)

**Comment by baywet** [comment]
**Also, for anyone reading this thread who would like the feature implemented and has not already upvoted [the issue first message](https://github.com/microsoft/vscode/issues/26289#issue-227317267), please do so as it influences prioritization for that team.**

**Comment by baywet** [comment]
@aeschli replying to your questions on [the PR here](https://github.com/microsoft/vscode/pull/327628#issuecomment-5438666955)

> I find it difficult to understand what the PR exactly tries to achieve.
> 
> Can we start with an issue that describes the requirements? I'm happy to take it from there and create PRs we don't need to discuss the details on how where what exactly be implemented (in the client, on the server, in the service...)
> 
> * you want to use schema associations from catalogs such as https://www.schemastore.org/api/json/catalog.json
>   
>   * is https://www.schemastore.org/api/json/catalog.json hardcoded or configurable
> * describe when these associations should be used
>   
>   * only when there is no other user or extension provided association
>   * is there a mechanism to disable certain associations
> * should schemas from the catalog be trusted or not
> * when should the catalog be reloaded


## Experience

I think the expected experience is the following for most users:
1. I open any JSON document in VSCode, if I hadn't configured specific schemas mapping in the JSON extension, or selected one through the selection tool, the following happens.
2. (on the first time I use the schema store, maybe a request to trust it)
3. Based on file naming patterns, or specific properties, a JSON Schema gets selected from the schema store
4. (if the selected entry from the schema store contains external references, get prompted to trust the sources of those external references, maybe on a per FQDN base to avoid getting too much noise)
5. I get structural validation, and potentially auto-completion and "hover labels" on properties based on the selected schema (those last two features are orthogonal to the schema source, just laying out the scenario here).

This is typically similar to the experience Visual Studio offers today for JSON.

## Caching

### Catalog and collocated schemas

In terms of caching behaviour, the catalog itself, or subsequent schemas that are collocated with the catalog, [gets about 20 PRs merged a week](https://github.com/SchemaStore/schemastore/pulls?q=is%3Apr+is%3Aclosed), so I think it'd make sense to cache it with a default or 4 hours or so? And pull it again when a JSON document is opened AND the catalog needs to be queried AND the cached version is older than the setting. Or something along these lines, you know the performance aspects of JSON experience in VSCode better than I do.

### External references

The frequency of updates to external references may vary widely based on the target. Some will remain stable for years on end, others might get updated on a daily basis (think of upcoming version of a standard that's not stable yet). So I think it'd make sense to align with the catalog caching itself for consistency? But again, you know the space better than I do.

## Configuration

The catalog structure itself doesn't have a specification (like an RFC or so), [but it does have a schema](https://www.schemastore.org/schema-catalog.json). I've never encountered additional catalogs in my experience, but I suspect some people might be using them? It also brings a host of questions like what if auth is required? etc... So I think in principle we can make the URL a defaulted setting, I'm not sure how much it'll be used in practice.

And yes, we might want to exclude some entries from the schema store for security reasons, or because they are noisy/poor quality. So an exclusion list seems like a good idea. I *think* this could be based on the name of the entry, which I *believe* is meant to be unique. TBD.

I hope that answers all your questions. Let me know if you have any additional comments or questions. Happy to follow up.

**Comment by madskristensen** [comment]
SchemaStore.org maintainer here. Please use <https://www.schemastore.org/api/json/catalog.json> as the URL if you consider moving this into VS Code.
